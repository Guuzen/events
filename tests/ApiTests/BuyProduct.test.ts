"use strict";
const TestDatabase = require('./Infrastructure/TestDatabase');
const testDatabase = new TestDatabase();
const Manager = require('./actors/Manager');
const Visitor = require('./actors/Visitor');
const Fondy = require('./actors/Fondy');

const visitorEmailServer = require('mockttp').getLocal();

jest.setTimeout(10 * 1000);

describe('Buy ticket', function () {

    beforeEach(async () => {
        await testDatabase.prepare();
        await visitorEmailServer.start(8001);
    });

    afterEach(async () => {
        await visitorEmailServer.stop();
    });

    test('By wire', async () => {
        const manager = new Manager();

        const eventId = await manager.createEventDomain();
        await manager.getEventsInList();
        await manager.getEventById(eventId);

        const tariffId = await manager.createTariff(eventId);
        await manager.getTariffsList(eventId);
        await manager.getTariffById(tariffId);
        await manager.createTariffDescription(tariffId);
        await manager.getTariffDescriptionById(tariffId);

        await manager.createFixedPromocode(eventId, tariffId);
        await manager.getPromocodeList(eventId);

        const visitor = new Visitor();
        const orderId = await visitor.placeOrder(tariffId);

        await manager.getOrdersList(eventId);
        await manager.getOrderById(orderId);

        await visitor.usePromocode(orderId, tariffId);
        await visitor.awaitsEmailWithTicket(visitorEmailServer);

        await manager.wait(1300);
        await manager.getOrderListWithUsedPromocode(eventId);
        await manager.getOrderByIdWithUsedPromocode(orderId);
        await manager.getPromocodeListWithUsedPromocode(eventId);

        await manager.markOrderPaid(eventId, orderId);

        await visitor.wait(1300);
        await visitor.receivesEmailWithTicket();

        await manager.getOrderListWithPaidOrder(eventId);
        await manager.getOrderByIdWithPaidOrder(orderId);
        const ticketId = await manager.getTicketsList(eventId);
        await manager.getTicketById(ticketId);
    })

    test('By card', async () => {
        const manager = new Manager();

        const eventId = await manager.createEventDomain();
        await manager.getEventsInList();
        await manager.getEventById(eventId);

        const tariffId = await manager.createTariff(eventId);
        await manager.getTariffsList(eventId);
        await manager.getTariffById(tariffId);
        await manager.createTariffDescription(tariffId);
        await manager.getTariffDescriptionById(tariffId);

        await manager.createFixedPromocode(eventId, tariffId);
        await manager.getPromocodeList(eventId);

        const visitor = new Visitor();
        const orderId = await visitor.placeOrder(tariffId);

        await manager.getOrdersList(eventId);
        await manager.getOrderById(orderId);

        await visitor.usePromocode(orderId, tariffId);

        await manager.wait(1300);
        await manager.getOrderListWithUsedPromocode(eventId);
        await manager.getOrderByIdWithUsedPromocode(orderId);
        await manager.getPromocodeListWithUsedPromocode(eventId);

        await visitor.payOrderByCard(orderId);
        await visitor.awaitsEmailWithTicket(visitorEmailServer);

        const fondy = new Fondy();
        await fondy.markOrderPaid(orderId);

        await visitor.wait(1300);
        await visitor.receivesEmailWithTicket();

        await manager.getOrderListWithPaidOrder(eventId);
        await manager.getOrderByIdWithPaidOrder(orderId);
        const ticketId = await manager.getTicketsList(eventId);
        await manager.getTicketById(ticketId);
    });
});
