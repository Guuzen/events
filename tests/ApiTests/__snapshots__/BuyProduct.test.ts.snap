// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Buy ticket By card: get event by id 1`] = `
Object {
  "data": Object {
    "domain": "localhost",
    "id": Any<String>,
    "name": "2019 foo event",
  },
}
`;

exports[`Buy ticket By card: get events in list 1`] = `
Object {
  "data": Array [
    Object {
      "domain": "localhost",
      "id": Any<String>,
      "name": "2019 foo event",
    },
  ],
}
`;

exports[`Buy ticket By card: get order by id 1`] = `
Object {
  "data": Object {
    "cancelled": false,
    "eventId": Any<String>,
    "id": Any<String>,
    "makedAt": Any<String>,
    "paid": false,
    "price": Object {
      "amount": "200",
      "currency": "RUB",
    },
    "productType": "ticket",
    "promocode": null,
    "promocodeId": null,
    "tariffId": Any<String>,
    "userId": Any<String>,
  },
}
`;

exports[`Buy ticket By card: get order by id with paid order 1`] = `
Object {
  "data": Object {
    "cancelled": false,
    "eventId": Any<String>,
    "id": Any<String>,
    "makedAt": Any<String>,
    "paid": true,
    "price": Object {
      "amount": "200",
      "currency": "RUB",
    },
    "productType": "ticket",
    "promocode": Object {
      "discount": Object {
        "amount": Object {
          "amount": "100",
          "currency": "RUB",
        },
        "type": "fixed",
      },
      "id": Any<String>,
    },
    "promocodeId": Any<String>,
    "tariffId": Any<String>,
    "userId": Any<String>,
  },
}
`;

exports[`Buy ticket By card: get order by id with used promocode 1`] = `
Object {
  "data": Object {
    "cancelled": false,
    "eventId": Any<String>,
    "id": Any<String>,
    "makedAt": Any<String>,
    "paid": false,
    "price": Object {
      "amount": "200",
      "currency": "RUB",
    },
    "productType": "ticket",
    "promocode": Object {
      "discount": Object {
        "amount": Object {
          "amount": "100",
          "currency": "RUB",
        },
        "type": "fixed",
      },
      "id": Any<String>,
    },
    "promocodeId": Any<String>,
    "tariffId": Any<String>,
    "userId": Any<String>,
  },
}
`;

exports[`Buy ticket By card: get order list with paid order 1`] = `
Object {
  "data": Array [
    Object {
      "cancelled": false,
      "eventId": Any<String>,
      "id": Any<String>,
      "makedAt": Any<String>,
      "paid": true,
      "price": Object {
        "amount": "200",
        "currency": "RUB",
      },
      "productType": "ticket",
      "promocode": Object {
        "discount": Object {
          "amount": Object {
            "amount": "100",
            "currency": "RUB",
          },
          "type": "fixed",
        },
        "id": Any<String>,
      },
      "promocodeId": Any<String>,
      "tariffId": Any<String>,
      "userId": Any<String>,
    },
  ],
}
`;

exports[`Buy ticket By card: get order list with used promocode 1`] = `
Object {
  "data": Array [
    Object {
      "cancelled": false,
      "eventId": Any<String>,
      "id": Any<String>,
      "makedAt": Any<String>,
      "paid": false,
      "price": Object {
        "amount": "200",
        "currency": "RUB",
      },
      "productType": "ticket",
      "promocode": Object {
        "discount": Object {
          "amount": Object {
            "amount": "100",
            "currency": "RUB",
          },
          "type": "fixed",
        },
        "id": Any<String>,
      },
      "promocodeId": Any<String>,
      "tariffId": Any<String>,
      "userId": Any<String>,
    },
  ],
}
`;

exports[`Buy ticket By card: get orders list 1`] = `
Object {
  "data": Array [
    Object {
      "cancelled": false,
      "eventId": Any<String>,
      "id": Any<String>,
      "makedAt": Any<String>,
      "paid": false,
      "price": Object {
        "amount": "200",
        "currency": "RUB",
      },
      "productType": "ticket",
      "promocode": null,
      "promocodeId": null,
      "tariffId": Any<String>,
      "userId": Any<String>,
    },
  ],
}
`;

exports[`Buy ticket By card: get promocode list 1`] = `
Object {
  "data": Array [
    Object {
      "allowedTariffs": Object {
        "tariffIds": Array [
          Any<String>,
        ],
        "type": "specific",
      },
      "code": "FOO",
      "discount": Object {
        "amount": Object {
          "amount": "100",
          "currency": "RUB",
        },
        "type": "fixed",
      },
      "eventId": Any<String>,
      "expireAt": "3000-01-01T12:00:00+00:00",
      "id": Any<String>,
      "usable": true,
      "useLimit": 10,
      "usedInOrders": Array [],
    },
  ],
}
`;

exports[`Buy ticket By card: get promocode list with used promocode 1`] = `
Object {
  "data": Array [
    Object {
      "allowedTariffs": Object {
        "tariffIds": Array [
          Any<String>,
        ],
        "type": "specific",
      },
      "code": "FOO",
      "discount": Object {
        "amount": Object {
          "amount": "100",
          "currency": "RUB",
        },
        "type": "fixed",
      },
      "eventId": Any<String>,
      "expireAt": "3000-01-01T12:00:00+00:00",
      "id": Any<String>,
      "usable": true,
      "useLimit": 10,
      "usedInOrders": Array [
        Any<String>,
      ],
    },
  ],
}
`;

exports[`Buy ticket By card: get tariff by id 1`] = `
Object {
  "data": Object {
    "eventId": Any<String>,
    "id": Any<String>,
    "priceNet": Array [
      Object {
        "price": Object {
          "amount": "200",
          "currency": "RUB",
        },
        "term": Object {
          "end": "3000-01-01T12:00:00+00:00",
          "start": "2000-01-01T12:00:00+00:00",
        },
      },
    ],
    "productType": "ticket",
  },
}
`;

exports[`Buy ticket By card: get tariff list 1`] = `
Object {
  "data": Array [
    Object {
      "eventId": Any<String>,
      "id": Any<String>,
      "priceNet": Array [
        Object {
          "price": Object {
            "amount": "200",
            "currency": "RUB",
          },
          "term": Object {
            "end": "3000-01-01T12:00:00+00:00",
            "start": "2000-01-01T12:00:00+00:00",
          },
        },
      ],
      "productType": "ticket",
    },
  ],
}
`;

exports[`Buy ticket By card: get ticket by id 1`] = `
Object {
  "data": Object {
    "createdAt": Any<String>,
    "eventId": Any<String>,
    "id": Any<String>,
    "number": Any<String>,
    "orderId": Any<String>,
  },
}
`;

exports[`Buy ticket By card: get tickets list 1`] = `
Object {
  "data": Array [
    Object {
      "createdAt": Any<String>,
      "eventId": Any<String>,
      "id": Any<String>,
      "number": Any<String>,
      "orderId": Any<String>,
    },
  ],
}
`;

exports[`Buy ticket By wire: get event by id 1`] = `
Object {
  "data": Object {
    "domain": "localhost",
    "id": Any<String>,
    "name": "2019 foo event",
  },
}
`;

exports[`Buy ticket By wire: get events in list 1`] = `
Object {
  "data": Array [
    Object {
      "domain": "localhost",
      "id": Any<String>,
      "name": "2019 foo event",
    },
  ],
}
`;

exports[`Buy ticket By wire: get order by id 1`] = `
Object {
  "data": Object {
    "cancelled": false,
    "eventId": Any<String>,
    "id": Any<String>,
    "makedAt": Any<String>,
    "paid": false,
    "price": Object {
      "amount": "200",
      "currency": "RUB",
    },
    "productType": "ticket",
    "promocode": null,
    "promocodeId": null,
    "tariffId": Any<String>,
    "userId": Any<String>,
  },
}
`;

exports[`Buy ticket By wire: get order by id with paid order 1`] = `
Object {
  "data": Object {
    "cancelled": false,
    "eventId": Any<String>,
    "id": Any<String>,
    "makedAt": Any<String>,
    "paid": true,
    "price": Object {
      "amount": "200",
      "currency": "RUB",
    },
    "productType": "ticket",
    "promocode": Object {
      "discount": Object {
        "amount": Object {
          "amount": "100",
          "currency": "RUB",
        },
        "type": "fixed",
      },
      "id": Any<String>,
    },
    "promocodeId": Any<String>,
    "tariffId": Any<String>,
    "userId": Any<String>,
  },
}
`;

exports[`Buy ticket By wire: get order by id with used promocode 1`] = `
Object {
  "data": Object {
    "cancelled": false,
    "eventId": Any<String>,
    "id": Any<String>,
    "makedAt": Any<String>,
    "paid": false,
    "price": Object {
      "amount": "200",
      "currency": "RUB",
    },
    "productType": "ticket",
    "promocode": Object {
      "discount": Object {
        "amount": Object {
          "amount": "100",
          "currency": "RUB",
        },
        "type": "fixed",
      },
      "id": Any<String>,
    },
    "promocodeId": Any<String>,
    "tariffId": Any<String>,
    "userId": Any<String>,
  },
}
`;

exports[`Buy ticket By wire: get order list with paid order 1`] = `
Object {
  "data": Array [
    Object {
      "cancelled": false,
      "eventId": Any<String>,
      "id": Any<String>,
      "makedAt": Any<String>,
      "paid": true,
      "price": Object {
        "amount": "200",
        "currency": "RUB",
      },
      "productType": "ticket",
      "promocode": Object {
        "discount": Object {
          "amount": Object {
            "amount": "100",
            "currency": "RUB",
          },
          "type": "fixed",
        },
        "id": Any<String>,
      },
      "promocodeId": Any<String>,
      "tariffId": Any<String>,
      "userId": Any<String>,
    },
  ],
}
`;

exports[`Buy ticket By wire: get order list with used promocode 1`] = `
Object {
  "data": Array [
    Object {
      "cancelled": false,
      "eventId": Any<String>,
      "id": Any<String>,
      "makedAt": Any<String>,
      "paid": false,
      "price": Object {
        "amount": "200",
        "currency": "RUB",
      },
      "productType": "ticket",
      "promocode": Object {
        "discount": Object {
          "amount": Object {
            "amount": "100",
            "currency": "RUB",
          },
          "type": "fixed",
        },
        "id": Any<String>,
      },
      "promocodeId": Any<String>,
      "tariffId": Any<String>,
      "userId": Any<String>,
    },
  ],
}
`;

exports[`Buy ticket By wire: get orders list 1`] = `
Object {
  "data": Array [
    Object {
      "cancelled": false,
      "eventId": Any<String>,
      "id": Any<String>,
      "makedAt": Any<String>,
      "paid": false,
      "price": Object {
        "amount": "200",
        "currency": "RUB",
      },
      "productType": "ticket",
      "promocode": null,
      "promocodeId": null,
      "tariffId": Any<String>,
      "userId": Any<String>,
    },
  ],
}
`;

exports[`Buy ticket By wire: get promocode list 1`] = `
Object {
  "data": Array [
    Object {
      "allowedTariffs": Object {
        "tariffIds": Array [
          Any<String>,
        ],
        "type": "specific",
      },
      "code": "FOO",
      "discount": Object {
        "amount": Object {
          "amount": "100",
          "currency": "RUB",
        },
        "type": "fixed",
      },
      "eventId": Any<String>,
      "expireAt": "3000-01-01T12:00:00+00:00",
      "id": Any<String>,
      "usable": true,
      "useLimit": 10,
      "usedInOrders": Array [],
    },
  ],
}
`;

exports[`Buy ticket By wire: get promocode list with used promocode 1`] = `
Object {
  "data": Array [
    Object {
      "allowedTariffs": Object {
        "tariffIds": Array [
          Any<String>,
        ],
        "type": "specific",
      },
      "code": "FOO",
      "discount": Object {
        "amount": Object {
          "amount": "100",
          "currency": "RUB",
        },
        "type": "fixed",
      },
      "eventId": Any<String>,
      "expireAt": "3000-01-01T12:00:00+00:00",
      "id": Any<String>,
      "usable": true,
      "useLimit": 10,
      "usedInOrders": Array [
        Any<String>,
      ],
    },
  ],
}
`;

exports[`Buy ticket By wire: get tariff by id 1`] = `
Object {
  "data": Object {
    "eventId": Any<String>,
    "id": Any<String>,
    "priceNet": Array [
      Object {
        "price": Object {
          "amount": "200",
          "currency": "RUB",
        },
        "term": Object {
          "end": "3000-01-01T12:00:00+00:00",
          "start": "2000-01-01T12:00:00+00:00",
        },
      },
    ],
    "productType": "ticket",
  },
}
`;

exports[`Buy ticket By wire: get tariff list 1`] = `
Object {
  "data": Array [
    Object {
      "eventId": Any<String>,
      "id": Any<String>,
      "priceNet": Array [
        Object {
          "price": Object {
            "amount": "200",
            "currency": "RUB",
          },
          "term": Object {
            "end": "3000-01-01T12:00:00+00:00",
            "start": "2000-01-01T12:00:00+00:00",
          },
        },
      ],
      "productType": "ticket",
    },
  ],
}
`;

exports[`Buy ticket By wire: get ticket by id 1`] = `
Object {
  "data": Object {
    "createdAt": Any<String>,
    "eventId": Any<String>,
    "id": Any<String>,
    "number": Any<String>,
    "orderId": Any<String>,
  },
}
`;

exports[`Buy ticket By wire: get tickets list 1`] = `
Object {
  "data": Array [
    Object {
      "createdAt": Any<String>,
      "eventId": Any<String>,
      "id": Any<String>,
      "number": Any<String>,
      "orderId": Any<String>,
    },
  ],
}
`;
